File: sdk/python/ppussh/_http.py
# ppussh/_http.py
"""
Shared async HTTP transport layer for the PPUSSH SDK.

Responsibilities
----------------
- Owns the single ``httpx.AsyncClient`` instance for each base URL.
- Implements the retry policy:
    * 5xx errors  → up to MAX_RETRIES attempts, exponential back-off (0.5s, 1s, 2s).
    * 429         → respects ``Retry-After`` header; falls back to RETRY_AFTER_DEFAULT.
    * 4xx (≠ 429) → never retried; raised immediately.
    * Network / timeout → treated like 5xx for retry purposes.
- Raises typed ``PpusshError`` subclasses — callers never see raw ``httpx`` exceptions.
"""
from __future__ import annotations

import asyncio
import logging
from typing import Any

import httpx

from ppussh.errors import PpusshAuthError, PpusshNetworkError, PpusshPaymentError, PpusshError

logger = logging.getLogger(__name__)

# ── Retry constants ────────────────────────────────────────────────────────────
MAX_RETRIES = 3
BACKOFF_SCHEDULE = [0.5, 1.0, 2.0]  # seconds between attempts 1→2, 2→3, 3→fail
RETRY_AFTER_DEFAULT = 1.0            # fallback when Retry-After header is absent
MAX_RATE_LIMIT_RETRIES = 2

# ── Timeout ────────────────────────────────────────────────────────────────────
DEFAULT_TIMEOUT = httpx.Timeout(connect=5.0, read=30.0, write=10.0, pool=5.0)

# ── SDK user-agent ─────────────────────────────────────────────────────────────
SDK_USER_AGENT = "ppussh-python/1.0.0"


class HttpTransport:
    """
    Thin wrapper around ``httpx.AsyncClient`` with retry logic baked in.

    One instance is created per base URL (accounts + payments) and shared
    across all method calls for the lifetime of the ``PpusshClient``.
    """

    def __init__(self, base_url: str) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = httpx.AsyncClient(
            base_url=self._base_url,
            timeout=DEFAULT_TIMEOUT,
            headers={
                "User-Agent": SDK_USER_AGENT,
                "Accept": "application/json",
            },
            follow_redirects=False,
        )

    # ── Public interface ───────────────────────────────────────────────────────

    async def request(
        self,
        method: str,
        path: str,
        *,
        headers: dict[str, str] | None = None,
        json: Any = None,
        data: dict[str, str] | None = None,   # form-encoded bodies
        params: dict[str, Any] | None = None,
        is_payments: bool = False,
    ) -> httpx.Response:
        """
        Execute an HTTP request with automatic retries.

        Parameters
        ----------
        method:       HTTP verb ("GET", "POST", etc.)
        path:         URL path relative to the base URL (e.g. ``"/oauth/token"``)
        headers:      Additional headers merged on top of the instance defaults.
        json:         JSON-serialisable body (sets Content-Type: application/json).
        data:         Form-encoded body dict (sets Content-Type: application/x-www-form-urlencoded).
        params:       Query string parameters.
        is_payments:  When True, payment-specific error parsing is used for 4xx responses.
        """
        kwargs: dict[str, Any] = {"params": params}
        if headers:
            kwargs["headers"] = headers
        if json is not None:
            kwargs["json"] = json
        if data is not None:
            kwargs["data"] = data

        attempt = 0
        rate_limit_attempt = 0
        last_exc: Exception | None = None

        while attempt < MAX_RETRIES:
            try:
                response = await self._client.request(method, path, **kwargs)
            except (httpx.ConnectError, httpx.ReadTimeout, httpx.WriteTimeout,
                    httpx.PoolTimeout, httpx.RemoteProtocolError) as exc:
                last_exc = exc
                attempt += 1
                if attempt < MAX_RETRIES:
                    delay = BACKOFF_SCHEDULE[min(attempt - 1, len(BACKOFF_SCHEDULE) - 1)]
                    logger.warning(
                        "ppussh: network error on %s %s (attempt %d/%d), "
                        "retrying in %.1fs: %s",
                        method, path, attempt, MAX_RETRIES, delay, exc,
                    )
                    await asyncio.sleep(delay)
                continue

            # ── 429 Rate limited ───────────────────────────────────────────────
            if response.status_code == 429:
                if rate_limit_attempt >= MAX_RATE_LIMIT_RETRIES:
                    raise PpusshNetworkError(
                        f"Rate limited on {method} {path} after "
                        f"{rate_limit_attempt + 1} attempts.",
                        status_code=429,
                        response_body=_safe_json(response),
                    )
                retry_after = _parse_retry_after(response)
                logger.warning(
                    "ppussh: rate limited on %s %s, retrying in %.1fs",
                    method, path, retry_after,
                )
                await asyncio.sleep(retry_after)
                rate_limit_attempt += 1
                continue  # does NOT consume an attempt from MAX_RETRIES

            # ── 5xx Server error ───────────────────────────────────────────────
            if response.status_code >= 500:
                attempt += 1
                if attempt < MAX_RETRIES:
                    delay = BACKOFF_SCHEDULE[min(attempt - 1, len(BACKOFF_SCHEDULE) - 1)]
                    logger.warning(
                        "ppussh: server error %d on %s %s (attempt %d/%d), "
                        "retrying in %.1fs",
                        response.status_code, method, path,
                        attempt, MAX_RETRIES, delay,
                    )
                    await asyncio.sleep(delay)
                    continue
                # All retries exhausted
                raise PpusshNetworkError(
                    f"Server error {response.status_code} on {method} {path} "
                    f"after {MAX_RETRIES} attempts.",
                    status_code=response.status_code,
                    response_body=_safe_json(response),
                )

            # ── 4xx Client errors (not retried) ───────────────────────────────
            if response.status_code >= 400:
                _raise_client_error(response, is_payments=is_payments)

            # ── 2xx / 3xx success ─────────────────────────────────────────────
            return response

        # Fell through MAX_RETRIES via network errors
        raise PpusshNetworkError(
            f"All {MAX_RETRIES} attempts failed for {method} {path}.",
            status_code=None,
        ) from last_exc

    async def aclose(self) -> None:
        """Close the underlying httpx client. Call when done with the SDK."""
        await self._client.aclose()

    # ── Context-manager support ────────────────────────────────────────────────

    async def __aenter__(self) -> HttpTransport:
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.aclose()


# ── Helpers ────────────────────────────────────────────────────────────────────

def _safe_json(response: httpx.Response) -> Any:
    try:
        return response.json()
    except Exception:
        return response.text


def _parse_retry_after(response: httpx.Response) -> float:
    header = response.headers.get("Retry-After")
    if header:
        try:
            return float(header)
        except ValueError:
            pass
    return RETRY_AFTER_DEFAULT


def _raise_client_error(response: httpx.Response, *, is_payments: bool) -> None:
    """Parse a 4xx response and raise the appropriate typed error."""
    body = _safe_json(response)

    # ── CONSENT_REQUIRED (403 structured body) ─────────────────────────────────
    if response.status_code == 403 and isinstance(body, dict):
        if body.get("status") == "CONSENT_REQUIRED":
            from ppussh.errors import PpusshConsentRequired
            raise PpusshConsentRequired(
                f"User has not consented to product '{body.get('product_name', '')}'.",
                client_id=body.get("client_id", ""),
                product_name=body.get("product_name", ""),
                product_description=body.get("product_description", ""),
                status_code=403,
                response_body=body,
            )

    # ── 401 Auth error ─────────────────────────────────────────────────────────
    if response.status_code == 401:
        detail = _extract_detail(body)
        raise PpusshAuthError(
            detail or f"Authentication failed (HTTP 401).",
            status_code=401,
            response_body=body,
        )

    # ── Payments service errors ────────────────────────────────────────────────
    if is_payments:
        code = body.get("code") if isinstance(body, dict) else None
        message = body.get("message") if isinstance(body, dict) else None
        raise PpusshPaymentError(
            message or f"Payments error (HTTP {response.status_code}).",
            code=code,
            status_code=response.status_code,
            response_body=body,
        )

    # ── Generic error ──────────────────────────────────────────────────────────
    detail = _extract_detail(body)
    raise PpusshError(
        detail or f"Request failed (HTTP {response.status_code}).",
        status_code=response.status_code,
        response_body=body,
    )


def _extract_detail(body: Any) -> str:
    """Pull the human-readable message out of an Accounts error envelope."""
    if isinstance(body, dict):
        return str(body.get("detail") or body.get("error") or body.get("message") or "")
    if isinstance(body, str):
        return body
    return ""

File: sdk/typescript/src/http.ts
// ppussh/src/http.ts
/**
 * Shared HTTP transport layer for the PPUSSH TypeScript SDK.
 *
 * Responsibilities:
 * - Owns an Axios instance per base URL.
 * - Implements the retry policy:
 *     5xx errors  → up to MAX_RETRIES attempts, exponential back-off (500ms, 1s, 2s).
 *     429         → respects Retry-After header; falls back to 1s default.
 *     4xx (≠ 429) → never retried; raised immediately.
 *     Network errors → treated like 5xx for retry purposes.
 * - Raises typed PpusshError subclasses; callers never see raw Axios errors.
 */

import axios, {
  AxiosInstance,
  AxiosRequestConfig,
  AxiosResponse,
  isAxiosError,
} from "axios";
import {
  PpusshAuthError,
  PpusshConsentRequired,
  PpusshError,
  PpusshNetworkError,
  PpusshPaymentError,
} from "./errors";

// ── Retry constants ──────────────────────────────────────────────────────────
const MAX_RETRIES = 3;
const BACKOFF_SCHEDULE = [500, 1000, 2000]; // ms
const RETRY_AFTER_DEFAULT = 1000; // ms fallback for missing Retry-After
const MAX_RATE_LIMIT_RETRIES = 2;

const SDK_USER_AGENT = "ppussh-ts/1.0.0";

// ── Helpers ──────────────────────────────────────────────────────────────────

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function parseRetryAfter(headers: Record<string, string | undefined>): number {
  const raw = headers["retry-after"] ?? headers["Retry-After"];
  if (raw !== undefined) {
    const secs = parseFloat(raw);
    if (!isNaN(secs)) return secs * 1000;
  }
  return RETRY_AFTER_DEFAULT;
}

function extractDetail(body: unknown): string {
  if (typeof body === "string") return body;
  if (body && typeof body === "object") {
    const b = body as Record<string, unknown>;
    return String(b["detail"] ?? b["error"] ?? b["message"] ?? "");
  }
  return "";
}

function raiseClientError(
  status: number,
  body: unknown,
  isPayments: boolean,
): never {
  // ── CONSENT_REQUIRED ───────────────────────────────────────────────────────
  if (
    status === 403 &&
    body &&
    typeof body === "object" &&
    (body as Record<string, unknown>)["status"] === "CONSENT_REQUIRED"
  ) {
    const b = body as Record<string, unknown>;
    throw new PpusshConsentRequired(
      `User has not consented to product '${b["product_name"] ?? ""}'.`,
      {
        clientId: String(b["client_id"] ?? ""),
        productName: String(b["product_name"] ?? ""),
        productDescription: String(b["product_description"] ?? ""),
        statusCode: 403,
        responseBody: body,
      },
    );
  }

  // ── 401 ───────────────────────────────────────────────────────────────────
  if (status === 401) {
    throw new PpusshAuthError(extractDetail(body) || "Authentication failed (HTTP 401).", {
      statusCode: 401,
      responseBody: body,
    });
  }

  // ── Payments errors ───────────────────────────────────────────────────────
  if (isPayments) {
    const b = body && typeof body === "object" ? (body as Record<string, unknown>) : {};
    throw new PpusshPaymentError(
      String(b["message"] ?? `Payments error (HTTP ${status}).`),
      {
        code: b["code"] ? String(b["code"]) : null,
        statusCode: status,
        responseBody: body,
      },
    );
  }

  // ── Generic ───────────────────────────────────────────────────────────────
  throw new PpusshError(extractDetail(body) || `Request failed (HTTP ${status}).`, {
    statusCode: status,
    responseBody: body,
  });
}

// ── RequestOptions ────────────────────────────────────────────────────────────

export interface RequestOptions {
  headers?: Record<string, string>;
  /** JSON body */
  json?: unknown;
  /** Form-encoded body (application/x-www-form-urlencoded) */
  form?: Record<string, string>;
  /** Query string parameters */
  params?: Record<string, string | number | boolean | undefined>;
  /** Set to true when calling the Payments service for correct error parsing */
  isPayments?: boolean;
}

// ── HttpTransport ─────────────────────────────────────────────────────────────

export class HttpTransport {
  private readonly _client: AxiosInstance;

  constructor(baseUrl: string) {
    this._client = axios.create({
      baseURL: baseUrl.replace(/\/$/, ""),
      timeout: 30_000,
      headers: {
        "User-Agent": SDK_USER_AGENT,
        Accept: "application/json",
      },
      // Never throw on non-2xx — we handle status codes ourselves
      validateStatus: () => true,
    });
  }

  async request(
    method: string,
    path: string,
    options: RequestOptions = {},
  ): Promise<AxiosResponse> {
    const { headers, json, form, params, isPayments = false } = options;

    const config: AxiosRequestConfig = {
      method,
      url: path,
      params,
      headers: { ...headers },
    };

    if (json !== undefined) {
      config.data = json;
      (config.headers as Record<string, string>)["Content-Type"] = "application/json";
    } else if (form !== undefined) {
      config.data = new URLSearchParams(form).toString();
      (config.headers as Record<string, string>)["Content-Type"] =
        "application/x-www-form-urlencoded";
    }

    let attempt = 0;
    let rateLimitAttempt = 0;
    let lastError: unknown = null;

    while (attempt < MAX_RETRIES) {
      let response: AxiosResponse;

      try {
        response = await this._client.request(config);
      } catch (err: unknown) {
        // Network / connection error
        lastError = err;
        attempt++;
        if (attempt < MAX_RETRIES) {
          const delay = BACKOFF_SCHEDULE[Math.min(attempt - 1, BACKOFF_SCHEDULE.length - 1)] ?? 2000;
          await sleep(delay);
        }
        continue;
      }

      const status = response.status;

      // ── 429 Rate limited ─────────────────────────────────────────────────
      if (status === 429) {
        if (rateLimitAttempt >= MAX_RATE_LIMIT_RETRIES) {
          throw new PpusshNetworkError(
            `Rate limited on ${method} ${path} after ${rateLimitAttempt + 1} attempts.`,
            { statusCode: 429, responseBody: response.data },
          );
        }
        const delay = parseRetryAfter(
          response.headers as Record<string, string | undefined>,
        );
        await sleep(delay);
        rateLimitAttempt++;
        continue; // does NOT consume from MAX_RETRIES
      }

      // ── 5xx Server error ─────────────────────────────────────────────────
      if (status >= 500) {
        attempt++;
        if (attempt < MAX_RETRIES) {
          const delay = BACKOFF_SCHEDULE[Math.min(attempt - 1, BACKOFF_SCHEDULE.length - 1)] ?? 2000;
          await sleep(delay);
          continue;
        }
        throw new PpusshNetworkError(
          `Server error ${status} on ${method} ${path} after ${MAX_RETRIES} attempts.`,
          { statusCode: status, responseBody: response.data },
        );
      }

      // ── 4xx Client errors (not retried) ──────────────────────────────────
      if (status >= 400) {
        raiseClientError(status, response.data, isPayments);
      }

      // ── 2xx success ───────────────────────────────────────────────────────
      return response;
    }

    // Fell through via network errors
    throw new PpusshNetworkError(
      `All ${MAX_RETRIES} attempts failed for ${method} ${path}.`,
      { cause: lastError },
    );
  }
}
